### DISCUSSION TITLE
Add mop and empty bin support to the vacuum component

### DISCUSSION THREAD

**Discussion by shred86** [home-assistant/architecture#684] [Entity Models]
## Context

Many robot vacuums (e.g. Roborock, Ecovacs, iRobot, etc.) have a mop and auto-empty bin feature with associated controls. For example, the Roborock S7 has the option to set the mop mode to standard or deep and adjust the mop scrub intensity to close (lifted), mild, moderate or intense. There is also the ability to start/stop the auto-emptying bin. These are not features or controls unique to a specific robot vacuum vendor or model.

This conversation started in issue #371. 

## Proposal

Add the following supported features to the base vacuum component:

- **Mop mode**: controls the routing for more thorough cleaning (i.e. overlap)
- **Mop intensity**: controls the scrub intensity of the mop
- **Water flow**: amount of water used while mopping
- **Empty bin**: start or stop the auto-emptying station when the robot vacuum is docked

## Consequences

Each integration implements the actual options for the supported feature (e.g. mop mode standard and deep). I do not see any consequences with adding these supported features.

**Comment by MikaBENOIT** [comment]
I'm interested in those features too.
It look like more and more vacuum tend to wash and scrub the floor too.
What are we waiting for a validation of this proposal ?

**Comment by Lash-L** [comment]
I think it makes sense for mopping functionality to potentially be a separate architecture issue than docking as they will both pose their own problems.

I'm going to only focus on mopping for right now. Mopping is being included in most mid-tier+ robot vacuums and there are even mop specific robot vacuums now, that just mop and don't vacuum. I don't think there is any issue with any integrations that wish to use a mop just implementing the Vacuum entity, but just something to keep in mind.

To add mopping, there are a few problems that would have to be solved
## Cleaning Actions
The Vacuum Entity should have a new CleaningAction attribute similar to climate's HVACAction
- MOPPING
- VACUUMING
- MOP_VACUUM? What is a good name for doing both at the same time?
- DOCKING? I know I said I was just focusing on mopping for right now, but this would probably make sense to be in the action
- OFF

## Cleaning Mode (Potentially these would just use the STATES that already exist?)
Similar to the actions, there should be new modes.
- OFF
- MOP
- VACUUM
- MOP_VACUUM
- DOCK

## VacuumEntityFeature
There should be new flags that can be enabled as a supported feature
- MOP_MODE
- MOP_INTENSITY
- MOP
- Vacuum? Right now it assumes the Vacuum entity can vacuum(which is logical) but if we add mopping functionality and we assume mop-only robot cleaners should use the Vacuum entity, I'm not sure how that would work.

## Current Functions
When we start the vacuum, we need to check state/ mode to determine what we should do. i.e. clean spot could either vacuum and mop or it could just mop or it could just vacuum. 

Does anyone have any further thoughts? I'd be happy to do a MVP/ initial implementation, but I'd be curious what core maintainers think. As the highest upvoted discussion post - there is obviously a high desire for this functionality.

**Comment by emontnemery** [comment]
The suggestion seems fine, but please list core integrations which could make use of the new features

**Reply by emontnemery** [reply]
@shred86 @MikaBENOIT @Lash-L Can you please list some brands of vacuum cleaners which could make use of this? It's preferred to list brands which already have a Home Assistant integration, but it's OK to also list brands which don't.

**Reply by Lash-L** [reply]
A few with mopping capabilities are
Roborock
Irobot
Ecovacs
dreamtech
Eufy

Most Robot vacuums are coming out with a mop even for the lower tiered versions now

**Reply by emontnemery** [reply]
Thanks, @Lash-L 👍

**Reply by CodeFatherG** [reply]
I am here searching for these features on a tuya vacuum, so Include tuya in the list

**Comment by emontnemery** [comment]
@Lash-L @shred86 the proposal is not clear enough to make a decision on.

To set the constraints a bit:
- The `VacuumEntity` class is deprecated and should not be considered in this proposal.
- The `StateVacuumEntity` is what should be considered to be updated

Please try to come up with a clear proposal outlining:
- New states, if any. The allowed states today are documented here: <https://developers.home-assistant.io/docs/core/entity/vacuum#states>
- New services, if any
- New feature flags, if any
- New state attributes, if any

**Comment by CodeFatherG** [comment]
I have a proposal to put forward. I am not as experienced at the auto empty docks but threw it in for completeness. I am taking an approach of rather than just cleaning or mopping, you leave it open ended for the integration to instead choose the mode to perform. IE, you can vacuum, or vacuum and mop, etc

- New states, if any. The allowed states today are documented here: https://developers.home-assistant.io/docs/core/entity/vacuum#states
  - STATE_CLEANING should remain to indicate that the vacuum is cleaning
    - However my proposal is that there should be an attribute to indicate the cleaning type. This allows integrations to implement their own types of clean; ie, vacuum, mop, vacuum and mop, deep clean, fast clean, etc
  - STATE_EMPTYING should be added to indicate the vacuum is in the action of emptying and not docking for charging
    - This would give state difference between STATE_RETURNING and STATE_EMPTYING
- New services, if any
  - set_cleaning_mode or async_set_cleaning_mode
    - Used to set the vacuum cleaning mode
  - empty or async_empty
    - Used to tell the vacuum to perform an empty at the dock
- New feature flags, if any
  - AUTO_EMPTY
    - Used to indicate this vacuum can perform emptying
- New state attributes, if any
  - cleaning_mode
    - Would indicate the cleaning action currently taking place
  - cleaning_modes
    - A list of cleaning modes for the vacuum
  - water_level
    - Some vacuums have the ability to set their water level independent of their cleaning modes.
  - some_variable_to_indicate_a_empty_required
    - I couldn't decide on a name, but this should indicate that the device needs an empty. Could be used for triggering an automation to notify the user when a manual empty is required (IE after 15 empties, empty the dock)

**Reply by CodeFatherG** [reply]
@emontnemery hi mate, been a long time since activity. I think it would be great to get this functionality off the ground given that now you can pick up this functionality for less than $400 AUD. It is so accessible now days that it feels like we are behind the curve on this one.

**Reply by CodeFatherG** [reply]
I have auto empty docks on the way for my vacuums so would love to get the ball rolling on this proposal and kick the discussion up again

**Comment by diedrichg** [comment]
I too would appreciate native control to empty the dust bin for those vacuums with a dock that automatically cleans the robot dust bin.

**Comment by sca075** [comment]
# Re-Opening Proposal: Standardised mop + auto-empty support in the `vacuum` integration

As previously discussed, a growing number of robot vacuums provide features that are not yet represented in the core `vacuum` entity model (e.g. mopping modes, hybrid mop+vacuum modes, and auto-empty docks). This proposal aims to:

1. Present a clear, concrete specification for additional feature flags and activity states.
2. Demonstrate current ecosystem demand across brands and integrations using Home Assistant Analytics (opt-in) data.

The intent is to keep full backwards compatibility while allowing integrations to expose modern capabilities in a consistent way.

---

## Ecosystem overview: vacuums using mop / auto-empty features

The following brands and integrations expose mopping, water-flow control, mop lifting, or auto-emptying via their APIs and/or custom components:

**Home Assistant Analytics (opt-in) – active installations**

Integration | Active installations
-- | --
Roborock | 32,774
iRobot Roomba / Braava | 10,823
dreame_vacuum (custom) | 7,965
xiaomi_cloud_map_extractor (custom) | 3,609
mqtt_vacuum_camera (custom) | 368
valetudo (custom) | 8

> These numbers are based on the Home Assistant Analytics opt-in data and therefore represent only a fraction of the actual user base. In practice, most new hybrid vacuums (even at the entry-level range) ship with a mop module and often an auto-empty dock.

This shows that tens of thousands of HA installations are already interacting with robot vacuums that support mop and/or auto-empty functionality today.

---

## Home Assistant integrations that can adopt these features

Core integrations that could immediately benefit from standardised mop/empty support include:

- [Roborock](https://www.home-assistant.io/integrations/roborock/)
- [Xiaomi Miio Vacuum](https://www.home-assistant.io/integrations/xiaomi_miio/#xiaomi-mi-robot-vacuum)
- [Ecovacs](https://www.home-assistant.io/integrations/ecovacs/)
- [iRobot / Roomba](https://www.home-assistant.io/integrations/roomba/)
- [MQTT Vacuum](https://www.home-assistant.io/integrations/vacuum.mqtt/)

Community integrations showing strong real-world demand:

- [Dreame Vacuums](https://github.com/Tasshack/dreame-vacuum)
- [Xiaomi Cloud Map Extractor](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor)
- [MQTT Vacuum Camera](https://github.com/sca075/mqtt_vacuum_camera)

Together, these integrations illustrate that mop modes, hybrid mop+vacuum modes, and auto-empty docks are now standard features rather than niche exceptions.

---

## Proposed new feature flags

The following additions extend `VacuumEntityFeature` to describe device capabilities. They do **not** change existing values or semantics.

> Names are illustrative; exact enum values/numbers can be adjusted to match core conventions.

Value | Description
-- | --
CLEAN_SPOT | The vacuum supports spot cleaning.
MOP | The vacuum supports mop cleaning (without vacuuming, or with very low suction).
MOP_VACUUM | The vacuum supports combined mop and vacuum cleaning modes.
AUTO_EMPTY | The vacuum supports self-emptying on a dock / station.
FAN_SPEED | The vacuum supports setting fan speed.
LOCATE | The vacuum supports locate / find-my-robot commands.
MAP | The vacuum supports retrieving its map.
PAUSE | The vacuum supports the pause command.
RETURN_HOME | The vacuum supports the return-to-dock command.
SEND_COMMAND | The vacuum supports sending arbitrary commands.
START | The vacuum supports the start command.
STATE | The vacuum supports returning its state.
STOP | The vacuum supports the stop command.

Existing features remain unchanged. The new flags are optional and only exposed by integrations that support them.

---

## Proposed extension of `VacuumActivity` states

Current implementations typically expose a single generic `cleaning` state. For modern hybrids, this loses important semantic information that is already available from the API (e.g. “mopping only”, “vacuum + mop”, “auto-emptying now”).

To keep compatibility while allowing more detail, we propose extending the `VacuumActivity` enum as follows:

```python
from enum import StrEnum


class VacuumActivity(StrEnum):
    """Vacuum activity states.

    These describe what the vacuum is currently doing, independent of brand
    or firmware. New values extend behaviour but do not change existing ones.
    """

    # Existing states – must remain unchanged for backward compatibility
    CLEANING = "cleaning"          # Generic cleaning when the mode is unknown
    DOCKED = "docked"
    IDLE = "idle"
    PAUSED = "paused"
    RETURNING = "returning"
    ERROR = "error"

    # New: more granular cleaning activities (for devices with MOP / MOP_VACUUM)
    VACUUMING = "vacuuming"        # Dry vacuum only
    MOPPING = "mopping"            # Mop only (no or minimal suction)
    VACUUMING_AND_MOPPING = "vacuuming_and_mopping"  # Hybrid mode

    # New: dock / station activity (for devices with AUTO_EMPTY)
    AUTO_EMPTYING = "auto_emptying"  # Dock is currently emptying the dust bin
```

```python
class VacuumEntityFeature(IntFlag):
    ...
    START = 8192

    MOP = 16384
    MOP_VACUUM = 32768
    AUTO_EMPTY = 65536
```

# In Summary

As proposed and discussed in #1303 with @gjohansson-ST and in the original architecture discussion:
https://github.com/home-assistant/architecture/discussions/684 @emontnemery @Lash-L @shred86 @diedrichg @CodeFatherG  the previous discussion was paused pending a clearer specification, so this proposal provides exactly that.

Home Assistant Analytics shows tens of thousands of vacuum integrations in active use, across multiple brands that already expose mop and auto-empty features via their APIs.

- The current vacuum model lacks a standard way to:

  * express the presence of mop / hybrid mop+vacuum / auto-empty capabilities, and
  
  * represent the corresponding runtime activities in a consistent, brand-agnostic way.

- The proposed additions:

  * Are fully backward compatible.
  
  * Keep existing states and features unchanged.
  
  * Provide a minimal but expressive set of new flags and activities that multiple core and custom integrations can adopt immediately.

I’d be happy to adjust naming, documentation, or mapping guidelines to better fit the expectations for the vacuum integration, but I hope this gives a clear, concrete basis for standardised mop and auto-empty support in Home Assistant core.

Created a **Ready for Review**: https://github.com/home-assistant/core/pull/156832 
The PR is updated including some change on the tests to cover the new states and also add the attributes discussed below.
Verified via Docker image of core + local integration: changes work end-to-end, no regressions detected.

The implementation follows the existing pattern used previously (i.e., same architecture, naming conventions, backwards-compatibility maintained).

Nothing else in the code base breaks because of this change — narrow, high-impact improvement.

**Reply by VIKTORVAV99** [reply]
I would just like to add that Matter is also a core integration that could benefit from this.

I have my robot vacuum connected via matter and it supports all of these proposed VacuumeActivity states (not 100% matter itself support all of these though).

**Reply by CodeFatherG** [reply]
Thanks for reopening this discussion! I'm glad to see progress on standardizing mop and auto-empty support. Overall the proposal looks solid, but I have some feedback on a few areas:

## Activity States

I don't see the value in adding a separate `VACUUMING` state when `CLEANING` already covers traditional vacuum operation. Are there specific use cases where distinguishing between `CLEANING` and `VACUUMING` provides meaningful value? Since there are no mop-only robot vacuums (they all have vacuum capability), I think we could simplify this to:
```python
class VacuumActivity(StrEnum):
    # Existing states – unchanged for backward compatibility
    CLEANING = "cleaning"          # Generic cleaning when mode is unknown
    DOCKED = "docked"
    IDLE = "idle"
    PAUSED = "paused"
    RETURNING = "returning"
    ERROR = "error"
    
    # New: more granular cleaning activities
    MOPPING = "mopping"            # Mop only (minimal/no suction)
    CLEANING_AND_MOPPING = "cleaning_and_mopping"  # Hybrid mode
    
    # New: dock activity
    AUTO_EMPTYING = "auto_emptying"  # Emptying dust bin at dock
```

## Incomplete Specification

The current spec defines feature flags and activity states but doesn't provide any way for users to actually interact with or control these new capabilities. Without corresponding services and attributes, the feature flags are essentially informational only. A complete specification needs to address how integrations would expose these controls and how users/automations would interact with them.

## Missing Services

The spec doesn't define services for controlling the new capabilities such as

- How do you **set the cleaning mode** for cleaning, mopping, and hybrid?
- How do you **set the water level** (fan level for mopping)?
- Trigger **empty dock**?
- Do you **start specific cleaning modes**? 
    - Start mopping
    - Start cleaning
    - Start Both

## Missing Attributes

Key attributes are missing that would be needed to expose device state and available options:

- **`cleaning_mode`** - Current active cleaning mode (vacuum, mop, hybrid, etc.)
- **`cleaning_modes`** - List of available cleaning modes for the device
- **`water_level`** - Current water flow setting (independent of cleaning mode)
- **`water_levels`** - Available water level options
- **`empty_required`** - Boolean indicating when dock needs manual emptying (useful for automation triggers)

Without corresponding [services/actions](https://github.com/home-assistant/core/blob/b3a1dd88c4cf913e8590de05578c9730bfb673a6/homeassistant/components/vacuum/__init__.py#L131) and [attributes/states](https://github.com/home-assistant/core/blob/b3a1dd88c4cf913e8590de05578c9730bfb673a6/homeassistant/components/vacuum/__init__.py#L375), the feature flags alone don't provide a complete interface.

Would you be open to expanding the proposal to include these missing pieces? I think addressing the service layer and state attributes would make this a more complete foundation for integrations to build on.

**Reply by sca075** [reply]
@CodeFatherG thanks,

Thanks for your feedback,  you’re right that the current proposal focuses on the representation layer (flags and states). I intentionally scoped it narrow to keep an additive, low-risk change.  Your follow-up proposal can cover the full services/attribute layer for mop/hybrid/auto-empty modes. The core team prefers probably a combined package.
What I can suggest:
  * Phase 1: Representation support (what we’ve submitted and expand also the attributes).
  * Phase 2: Templating the services.

What can easily add in the PR is:
```python
ATTR_BATTERY_ICON = "battery_icon"
ATTR_CLEANED_AREA = "cleaned_area"
ATTR_FAN_SPEED = "fan_speed"
ATTR_FAN_SPEED_LIST = "fan_speed_list"
ATTR_PARAMS = "params"
ATTR_STATUS = "status"
ATTR_CLEANING_MODE = "cleaning_mode"
ATTR_CLEANING_MODES_LIST = "cleaning_modes_list"
ATTR_WATER_LEVEL = "water_level"
ATTR_WATER_LEVELS_LIST = "water_levels_list"
ATTR_EMPTY_REQUIRED = "empty_required"
```

This can help the integrations authors to start to adapt them code and eventually create services as required at integrations level.

Another good point would be to add the standardised cleaning_modes I would go for this [Vacuuming, Mopping, Vacuuming and Mopping] and water_levels [Slightly dry, Moist, Wet]. One think to probably do is to add a thin logic in this template as per the water_levels can't be set when mode is just Vacuuming.

Hope this cover the most.

**Reply by VIKTORVAV99** [reply]
Could we add another dock activity as well called something along the line of `MOP_DRYING` or `DRYING_MOP` for when the robot is docked and drying it's mops?

Several different robots and brands such as Roborock, Roomba, and SwitchBot already supports this.

**Reply by sca075** [reply]
@VIKTORVAV99 sure I would also add the State "cleaning mop pads".

**Reply by VIKTORVAV99** [reply]
That works! But maybe skip the pads as some are using rollers now?

**Reply by sca075** [reply]
> That works! But maybe skip the pads as some are using rollers now?

yep some new models but for the most actually mop pads is more specific than "mopping tools" so, in any cases the PR I opened gives an overview.. I did try to keep it really simple, @CodeFatherG , did not change the start command for now but I added the new possible starts to the device_trigger.  Also some service were added to change cleaning mode and wet level... repeat it is a kind of "device template" it can probably be manage now  as we have the boolean attribute for emptying the dock.

**Comment by sca075** [comment]


**Reply by sca075** [reply]
When this will be approved, as the PR was closed as we miss the approval here. Just to be precise, the changes are ready and fully tested but some improvements can be done for the nomenclature of some of the things. The dead line is Feb of 2026 I guess (for instance in the code comments). 

For instance we could consider:

1) the cleaning and docked states are the main core design and I would propose to maintain those untouched.
2) mopping, vacuuming or combination of both can be a sub-states when the robot offers them an could be added to the automation triggers.
3) As some robots just.. mop this should be defined as it was proposed in the vacuum futures.
4) auto empty is a future of the dock station and at current is a beat ambiguous.. we need to define it a little better as per there is a also to clean sometimes the self_cleaning station (mops or bags to be replaced as well water level)
5) water level and wet level actually needs to be separated as tanks are in the self_cleaning stations and wetness of the mop are different and strictly part of mopping activity.

As per this in any cases is and must be a template I think also that makes sense to seriously implement also services to reset the counters. Most importantly agree the nomenclature of each thing.. at this point is the most important thing so will review few points and keep the current implementation available just in case. Whether any comment or feedback from the core team is available would be appreciated.

**Comment by incline02** [comment]
any news on this ?
